sw/host/sphincsplus/lib.rs

```diff
@@ -2,16 +2,149 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
-use sphincsplus_bindgen::SPX_BYTES;
+use sphincsplus_bindgen::crypto_sign_keypair;
+use sphincsplus_bindgen::crypto_sign_seed_keypair;
+use sphincsplus_bindgen::crypto_sign_signature;
+use sphincsplus_bindgen::crypto_sign_verify;
 
-pub fn do_something() -> u32 {
-    SPX_BYTES
+// SPHINCS+ secret key byte length.
+pub const SPX_SECRET_KEY_BYTES: usize = sphincsplus_bindgen::CRYPTO_SECRETKEYBYTES as usize;
+pub const SPX_PUBLIC_KEY_BYTES: usize = sphincsplus_bindgen::CRYPTO_PUBLICKEYBYTES as usize;
+pub const SPX_SIGNATURE_BYTES: usize = sphincsplus_bindgen::CRYPTO_BYTES as usize;
+pub const SPX_SEED_BYTES: usize = sphincsplus_bindgen::CRYPTO_SEEDBYTES as usize;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SpxPublicKey([u8; SPX_PUBLIC_KEY_BYTES]);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SpxSecretKey([u8; SPX_SECRET_KEY_BYTES]);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SpxSignature([u8; SPX_SIGNATURE_BYTES]);
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SpxSeed([u8; SPX_SEED_BYTES]);
+
+// Generate a new keypair from a seed.
+pub fn spx_keypair_from_seed(seed: &SpxSeed) -> anyhow::Result<(SpxPublicKey, SpxSecretKey)> {
+    let mut pk = [0u8; SPX_PUBLIC_KEY_BYTES];
+    let mut sk = [0u8; SPX_SECRET_KEY_BYTES];
+    let err_code =
+        // SAFETY: the buffers here are all fixed-length arrays of the size expected by the C code.
+        unsafe { crypto_sign_seed_keypair(pk.as_mut_ptr(), sk.as_mut_ptr(), seed.0.as_ptr()) };
+    anyhow::ensure!(
+        err_code == 0,
+        "SPHINCS+ seeded key pair generation failed with error code {err_code}."
+    );
+    Ok((SpxPublicKey(pk), SpxSecretKey(sk)))
+}
+
+// Generate a new random keypair.
+pub fn spx_keypair_generate() -> anyhow::Result<(SpxPublicKey, SpxSecretKey)> {
+    let mut pk = [0u8; SPX_PUBLIC_KEY_BYTES];
+    let mut sk = [0u8; SPX_SECRET_KEY_BYTES];
+    let err_code =
+        // SAFETY: the buffers here are all fixed-length arrays of the size expected by the C code.
+        unsafe { crypto_sign_keypair(pk.as_mut_ptr(), sk.as_mut_ptr()) };
+    anyhow::ensure!(
+        err_code == 0,
+        "SPHINCS+ random key pair generation failed with error code {err_code}."
+    );
```

Comment:
Please return structured errors from library functions.  The `anyhow` ensure/bail functions, when taking a string as the error argument, basically are returning a "stringly-typed" error.  Should a user of this library want to handle or otherwise make behavior decisions based on the error, stringly-typed errors then require string parsing.

My recommendation is to avoid the use of the `anyhow` crate in low-level base libraries.  You can use the `thiserror` crate to easily create nice error types and return those errors via `std::result::Result`.  Higher-level code is then free to handle such errors or thunk those errors into an `anyhow::Result` via `?` coercion.

Based on what I see in this PR, I'd perhaps start like this:
```
use thiserror::Error;

#[derive(Debug, Error)]
pub enum SpxError {
    #[error("Key generation failed with error {0}")]
    KeyGen(u32),

    #[error("Unexpected length {0}")]
    BadLength(u32),
}
```

Reply:
Fixed!